### PR TITLE
Propagating scope information in indirect application to a reference.

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** enhancement

The following shows two situations where known scopes of a reference are not propagated while it could:
```coq
Notation "0" := true : bool_scope.
Axiom f : bool -> bool -> nat.
Check (@f 0) 0.
(* The term "0" has type "nat" while it is expected to have type "bool". *)
Record R := { p : bool -> nat }.
Check fun r => r.(@p) 0.
(* The term "0" has type "nat" while it is expected to have type "bool". *)
```
The PR makes that this works.

- [X] Added / updated test-suite
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

File: interp/constrintern.ml
Comment (SkySkimmer) [inline interp/constrintern.ml:2055]: Please put the ending `in` of a multiline `in` on a separate line (I find it harder to understand the code structure otherwise).

Comment (herbelin) [reply interp/constrintern.ml:2055]: Ah, I always do like this... Could we then try to state an explicit recommendation about this point in [`dev/doc/styles.txt`](https://github.com/coq/coq/blob/33e748514dad9459885006a1523d107d556be22b/dev/doc/style.txt), that I can then follow consistently.

**Comment by herbelin** [comment]
This is a small change in a border case of interpretation scopes, which seems natural. If no developers has an opinion, maybe the best is to ask for user feedback on Zulip?

File: interp/constrintern.ml
Comment (herbelin) [reply interp/constrintern.ml]: @skyskimmer: I finally did as you suggested.

After more thinking, I believe that the change is natural enough to not need prior comments from users. So, looking for an @coq/extensible-syntax-maintainers's assignee (it is very short, so, it should not take more than 5 minutes).

**Comment by herbelin** [comment]
After all, the examples addressed with this PR are relatively border cases. No need to bother users for them. I believe the new proposed semantics is intuitive enough and w/o impact on the CI. So, it can be merged. Looking thus for an assignee (@SkySkimmer, would you accept?).

**Comment by ppedrot** [comment]
Despite your claim in the private communication that you sent to other cod devs yesterday, this is far from being trivial. Parentheses are significant in Gallina, although I never quite understood the motivation for that. So this qualifies as a design change and needs to be discussed with end-users.

**Comment by herbelin** [comment]
@ppedrot: Thanks for the comments! Is it correct that you have several questions:
1. is it relevant that `(@f 0) 0` behaves differently from `@f 0 0`?
2. is the design change (including the one in `fun r => r.(@p) 0`, which is independent from the previous item) worth?

I have no opinion on the first item except that it is a priori useful (but at the same time probably unused since CI did not fail).

About point 2, I was thinking about asking users at some time but then I felt that users would say that they don't have an opinion on an obscure feature they were not even aware of (again based on the CI not having nothing a change). I'm however ready to ask if you have an idea of how to proceed to ask.

**Comment by ppedrot** [comment]
> is it relevant that (@f 0) 0 behaves differently from @f 0 0?

Let's say I wouldn't be surprised if after the merge of this PR we see some power-user ask a question about the non-uniformity of handling of arguments w.r.t. scopes and implicits. Iris devs have a knack for this kind of involved behaviours.

> is the design change (including the one in fun r => r.(@p) 0, which is independent from the previous item) worth?

No idea but my personal belief is that we should stop messing with anything related to projections without settling for a decision about the `r.(p)` syntax. Last time I heard, there were still two possible solutions, namely 1. the introduction of a new syntax for primitive projections and 2. the deprecation of that syntax for non-primitive projections. Nothing has moved since.

> probably unused since CI did not fail

I agree that the change introduced in this PR is really quite hard to hit, but it doesn't prevent us from trying to be as careful as possible when changing behaviours.

**Comment by herbelin** [comment]
> No idea but my personal belief is that we should stop messing with anything related to projections without settling for a decision about the r.(p) syntax. Last time I heard, there were still two possible solutions, namely 1. the introduction of a new syntax for primitive projections and 2. the deprecation of that syntax for non-primitive projections. Nothing has moved since.

In my case, I'm working on many things at the same time, so without a ping of some form, the daily needs tend to come in front. The best thing to make progress is probably to work in collaboration with deadlines which force us to make progresses. We can do that if you want.

I think you know my opinion which is that I don't see a reason to make a difference between positive and negative projections regarding the `t.(p)` notation but, in any case, I agree that we should restrict it exclusively to (fully-applied) projections. If this direction is ok, we could say that I (or you if you prefer) do it for say the 16th of November.

The alternative would be to have two distinct (and respectively exclusive) syntaxes for positive and negative projections, but is there something to gain of two distinct syntaxes?

A question is also about parameters and implicit arguments. Do we accept e.g. to provide parameters to negative projections, even if it is not kept in the internal representation. That could be e.g. useful as a typing hint. In particular, do we accept that some parameters can be declared as explicit (even if later written `_` by the printer)?

**Comment by herbelin** [comment]
@ppedrot:
>  Last time I heard, there were still two possible solutions, namely 1. the introduction of a new syntax for primitive projections and 2. the deprecation of that syntax for non-primitive projections. Nothing has moved since.

So, is it then ok if we go for two syntaxes? After the 8.13 branching, I can concentrate on reviving #6651 if you want. Will you support it? As for the exact syntax, I guess there are plenty of possible choices, e.g. say, `t.(|p|)`, reminiscent of `{|` and `|}`.

**Comment by ppedrot** [comment]
> Will you support it?

I am not sure what kind of mental model you have of the Coq dev decision process, but let me state mine so as to be clear.

It's not a matter of me supporting something, or any other individual for that matter. Decisions need to be taken collegially, and with the interaction of enlightened power-users that actually happen to use the stuff we are haphazardly developing. Since unanimity is a fantasy only the EU believes in, this also means that there will be conflict between points of view that need to be resolved either by some kind of majority rule, an appeased balance of power, or, unfortunately more often in the Coq dev process, by "the last one standing has won". But at the very least non-trivial user-facing changes need to be discussed, especially with our users.

In particular, the choice between the two options for primitive projections above needs to be broadly advertised, because it it highly non-trivial. Then we can take a sovereign decision as developers, potentially going against the will of our users. But at least it will be an informed one.

**Comment by herbelin** [comment]
In the present case, I have no particular mental model of the Coq dev decision process in mind. It is rather that I'm willing that the decision processes does not stall too much. In particular, it seems to me to be a fact that we don't have enough manpower to all discuss about all decisions to take (even to at least one decision being discussed by at least 2 of us). The case of primitive projections is an example (#6651 is from 2018).

In the case of primitive projections, I developed a proposal in length without succeeding in getting feedback. I'm talking to you because you expressed opinions on primitive projections and I had the impression that at least the two of us can make progresses altogether. But I'm fine in any other form of Coq dev decision process which can reasonably be made convergent. For instance, can we target that within e.g. two months, one way or another, we will have resolved the pending primitive projection issues?

PS: among ways to be more efficient, I think that the kind of discussion we have at this point would work better by talking.

**Comment by herbelin** [comment]
> Parentheses are significant in Gallina, although I never quite understood the motivation for that. So this qualifies as a design change and needs to be discussed with end-users.

Actually, I should have phrased my former [comment](https://github.com/coq/coq/pull/12685#issuecomment-722488963) in past tense: "1. was it relevant that `(@f 0) 0` behaved differently from `@f 0 0`?". Indeed, and as far as I can judge, to the exception of significance of parentheses wrt scopes that this PR precisely "fixes", I can't find other examples where parentheses were significant in applicative terms. So, it seems to me that the `(@f 0) 0` part of the PR qualifies more as a fix than as a design change, don't you think?

Thinking further at your comments. I don't think that this PR is about "messing" with projections, and it is orthogonal per se to the question of primitive projections (even if I agree that we should eventualy settle a decision about the syntax of primitive projections). I defend that the taking into account of scopes in projection is a the "right", intuitive, expected thing to do, even in the presence of `@`. This is also consistent with #11120.

If you think that there is a risk that the change is considered unexpected but I may still post a request for comments on Zulip or Discourse, if you insist. Alternatively, I could also check with Iris dev, which are generally responsive.

In any case, if we could arrive to a conclusion on this PR by 8.13 without too much stress, that would make me happy.

-------
The rest of this comment is a side question

As for the `t.(p)` part, and independently of the PR, we can question whether it is expected that `@` of a (positive) projection deactivates not only the implicit parameters of the record but also the implicit arguments of the proper type of the projection:
```
Record R {m} (H:m=0) := { p : forall {n}, n=0 -> True }.
Check fun r => r.(p eq_refl) eq_refl.
Check fun r => r.(@p 0 eq_refl) 0 eq_refl.
Fail Check fun r => r.(@p 0 eq_refl) eq_refl.
```

This is however a side question which is not changed by the PR, as the PR is only about scopes.

**Comment by herbelin** [comment]
@ppedrot: I would warmly appreciate an answer from you before 8.13 is frozen! Even if it is to say e.g. that you don't have time to concentrate on it, or that you need a few days, or whatever reason, so that I know where we stand!

**Comment by ppedrot** [comment]
I don't have much time, but I don't have a strong opinion about this PR. As far as I am concerned it can go in, *but* once again it lacks motivation (this is an archetypical instance of the headless duck development strategy), and the current behaviour seems vaguely more uniform to me w.r.t. the semantics of parentheses in Gallina (but only vaguely so, so I am not taking sides either).

In any case, I don't really see why it should be considered urgent since it's really a fringe behaviour that virtually nobody relies on.

**Comment by ppedrot** [comment]
(We can have a more general discussion at the weekly call. It's happening this afternoon, right?)

**Comment by herbelin** [comment]
Your answer is multi-facets...

>  archetypical instance of the headless duck development strategy

What to do to address this point? Postponing to Greek calends does not seem better than "local stepwise fine-tuning" if I may formulate more positively your expression (not that I'm opposed to headed development strategy though, of course, and I'm sure we can do better on this point!).

> it lacks motivation

The motivation is intuitiveness and regularity (which I agree may be subjective, so either we try to make it objective, or we trust the corresponding "maintainer", or whatever solution finds agreement, but we need to be constructive one way or another).

> In any case, I don't really see why it should be considered urgent since it's really a fringe behaviour that virtually nobody relies on.

It is nothing urgent per se, but postponing would not help either.

> (We can have a more general discussion at the weekly call. It's happening this afternoon, right?)

Yes. Shall do.

**Comment by ppedrot** [comment]
The decision of the call was to merge as-is. @herbelin can you rebase to get a clean CI? I'll merge after if no issue.

**Review by ppedrot — APPROVED** [review]
Code change is fine.

**Pull request merged**
